OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalStateSynchronizer.java

```diff
@@ -174,9 +203,46 @@ static void logoutEmail() {
       getEmailStateSynchronizer().logoutEmail();
    }
 
-   static void setExternalUserId(String externalId) throws JSONException {
-      getPushStateSynchronizer().setExternalUserId(externalId);
-      getEmailStateSynchronizer().setExternalUserId(externalId);
+   static void setExternalUserId(String externalId, final OneSignal.OSExternalUserIdUpdateCompletionHandler completionHandler) throws JSONException {
+      final JSONObject responses = new JSONObject();
+      // Create a handler for internal usage, this is where the developers completion handler will be called,
+      //  which happens once all handlers for each channel have been processed
+      OneSignal.OSInternalExternalUserIdUpdateCompletionHandler handler = new OneSignal.OSInternalExternalUserIdUpdateCompletionHandler() {
+         @Override
+         public void onComplete(String channel, boolean success) {
+            OneSignal.onesignalLog(OneSignal.LOG_LEVEL.VERBOSE, "Completed request to update external user id for channel: " + channel + " and success: " + success);
+            try {
+               // Latest channel success status will be overwriting previous ones since we only care about latest request status
+               // That way the completion handler is allowing the developer to handle the most recent scenario
+               responses.put(channel, new JSONObject().put("success", success));
+            } catch (JSONException e) {
+               OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "Error while adding the success status of external id for channel: " + channel);
+               e.printStackTrace();
+            }
+
+            for (UserStateSynchronizer userStateSynchronizer : userStateSynchronizers.values()) {
+               if (userStateSynchronizer.hasQueuedHandlers()) {
```

Comment:
this that if for one userStateSynchronizer there are handler this will cancel the callback. Why isnt this called before the try catch?

Reply:
We check if any other handlers for the Push or Email channel exist. We wont return completion until all handlers are done being processed

Reply:
Steps until this point:
1. Init responses
2. Create handler
3.  Add handler to each channel
4. Await for completion and all handlers to be done on each channel
5. All channels done processing handlers, now lets return completion with most recent status from each channel

Reply:
oooh i see, i looked it wrong, Thx!